tss-esapi/src/abstraction/web_authn.rs

```diff
@@ -0,0 +1,187 @@
+use crate::{
+    constants::SessionType,
+    handles::{KeyHandle, SessionHandle},
+    interface_types::algorithm::HashingAlgorithm,
+    structures::{
+        pcr_selection_list::PcrSelectionListBuilder, Attest, HashScheme, Public, Signature,
+        SignatureScheme, SymmetricDefinition,
+    },
+    traits::Marshall,
+    Context, Result,
+};
+use ciborium::cbor;
+use std::convert::TryInto;
+
+#[derive(Debug)]
+pub struct TpmStatement {
+    attestation: Attest,
+    signature: Signature,
+    public_area: Public,
+    algorithm: SignatureScheme,
+    x509_chain: Vec<Vec<u8>>,
+}
+
+impl TpmStatement {
+    pub fn new(
+        context: &mut Context,
+        object: KeyHandle,
+        key: KeyHandle,
+        nonce: Vec<u8>,
+    ) -> Result<TpmStatement> {
+        let (public_area, _, _) = context.read_public(object)?;
+        let (session_1, _, _) = context.sessions();
+        let session_2 = context.start_auth_session(
+            None,
+            None,
+            None,
+            SessionType::Hmac,
+            SymmetricDefinition::AES_128_CFB,
+            HashingAlgorithm::Sha256,
+        )?;
+        let signing_scheme = SignatureScheme::RsaSsa {
+            hash_scheme: HashScheme::new(HashingAlgorithm::Sha256),
+        };
+        let (attestation, signature) = context
+            .execute_with_sessions((session_1, session_2, None), |ctx| {
+                ctx.certify(object.into(), key, nonce.try_into()?, signing_scheme)
+            })
+            .or_else(|e| {
+                context.flush_context(object.into())?;
+                context.flush_context(key.into())?;
+                context.flush_context(SessionHandle::from(session_2).into())?;
+                Err(e)
+            })?;
+        context.flush_context(object.into())?;
+        context.flush_context(key.into())?;
+        context.flush_context(SessionHandle::from(session_2).into())?;
+        Ok(TpmStatement {
+            attestation,
+            signature,
+            x509_chain: Vec::new(),
+            algorithm: signing_scheme,
+            public_area,
+        })
+    }
+
+    pub fn add_certificates(&mut self, mut certificates: Vec<Vec<u8>>) {
+        self.x509_chain.append(&mut certificates);
+    }
+
+    pub fn convert_to_tpmstmtformat_and_encode(&self) -> Vec<u8> {
+        let cert_info_value: Vec<u8>;
+        let sig_value: Vec<u8>;
+        let pub_area_value: Vec<u8>;
+        match self.signature.clone().marshall() {
+            Ok(sig) => sig_value = sig,
+            Err(_) => panic!("tried to convert Err type"),
```

Comment:
Hmm, I think it'd be better to return `Result<Vec<u8>>` from this function - panicking/unwrapping is something we try to avoid at all costs because it essentially crashes the user app. Then you can just `?` out all the `Result`s you get along the way